libcaf_core/caf/chrono.hpp

```diff
@@ -0,0 +1,159 @@
+  using clock_duration = typename clock_type::duration;
+  auto tse = ts.time_since_epoch();
+  clock_timestamp as_sys_time{sc::duration_cast<clock_duration>(tse)};
+  auto secs = clock_type::to_time_t(as_sys_time);
+  auto ns = sc::duration_cast<sc::nanoseconds>(tse).count() % 1'000'000'000;
+  return {secs, static_cast<int>(ns)};
+}
+
+/// Prints the date and time to a buffer.
+/// @param buf The buffer to write to.
+/// @param buf_size The size of `buf` in bytes.
+/// @param secs The number of seconds since epoch.
+/// @param nsecs The number of nanoseconds since the last full second.
+/// @returns A pointer to the null terminator of `buf`.
+/// @pre `buf_size >= 36`.
+CAF_CORE_EXPORT
+char* print_localtime(char* buf, size_t buf_size, time_t secs,
+                      int nsecs) noexcept;
+
+} // namespace caf::detail
+
+namespace caf::chrono {
+
+// TODO: becomes obsolete when switching to C++20.
+template <class Duration>
+using sys_time = std::chrono::time_point<std::chrono::system_clock, Duration>;
+
+/// Formats `ts` in ISO 8601 format.
+/// @param ts The time point to format.
+/// @returns The formatted string.
+template <class Duration>
+std::string to_string(sys_time<Duration> ts) {
+  char buf[detail::format_buffer_size];
+  auto [secs, nsecs] = detail::split_time_point(ts);
+  detail::print_localtime(buf, sizeof(buf), secs, nsecs);
+  return buf;
+}
+
+/// Prints `ts` to `out` in ISO 8601 format.
+/// @param out The buffer to write to.
+/// @param ts The time point to print.
+template <class Buffer, class Duration>
+void print(Buffer& out, sys_time<Duration> ts) {
+  char buf[detail::format_buffer_size];
+  auto [secs, nsecs] = detail::split_time_point(ts);
+  auto end = detail::print_localtime(buf, sizeof(buf), secs, nsecs);
+  out.insert(out.end(), buf, end);
+}
+
+/// Represents a point in time, expressed as a date and time of day. Also
+/// provides formatting and parsing functionality for ISO 8601.
+class CAF_CORE_EXPORT datetime {
+public:
+  /// The year.
+  int year = 0;
+
+  /// The month of the year, starting with 1 for January.
+  int month = 0;
+
+  /// The day of the month, starting with 1.
+  int day = 0;
+
+  /// The hour of the day, starting with 0.
+  int hour = 0;
+
+  /// The minute of the hour, starting with 0.
+  int minute = 0;
+
+  /// The second of the minute, starting with 0.
+  int second = 0;
+
+  /// The nanosecond of the second, starting with 0.
+  int nanosecond = 0;
+
+  /// The offset from UTC in seconds.
+  std::optional<int> utc_offset;
+
+  /// Returns whether this object contains a valid date and time. A default
+  /// constructed object is invalid.
+  bool valid() const noexcept;
+
+  /// Returns whether this object is equal to `other`.
+  bool equals(const datetime& other) const noexcept;
+
+  /// Parses a date and time string in ISO 8601 format.
+  /// @param str The string to parse.
+  /// @returns The parsed date and time.
+  error parse(std::string_view str);
```

Comment:
For the [IP address](https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/caf/ipv4_address.hpp#L131) and [UUID](https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/caf/uuid.hpp#L154) classes this is a free function, do you think it makes sense to have this be a member in this case?

Reply:
Uh, good thing that you caught that! 👍

Makes zero sense to have a member function here but a free function everywhere else. Fixed.